### PR TITLE
Remove redundant Contravariant constraint from cosmos[On,Of]

### PR DESCRIPTION
**Pull request by sjakobi**
@Gabriel439 had pointed this out.

### CODE REVIEW

**Comment by sjakobi** [comment]
Maybe it would be good to enable `-Wredundant-constraints` in the test suite?

**Comment by emilypi** [comment]
In practice this would be a `Fold`, so that `Contravariant` constraint on the `LensLike` is probably for better inference. Instead of removing the constraint, could you see what breaks when you monomorphise to a `Fold s a`

File: src/Control/Lens/Plated.hs [hunk 1]
```diff
@@ -444,7 +444,7 @@ cosmosOf d f s = f s *> d (cosmosOf d f) s
 -- @
 -- 'cosmosOn' :: 'Plated' a => 'Fold' s a -> 'Fold' s a
 -- @
-cosmosOn :: (Applicative f, Plated a) => LensLike' f s a -> LensLike' f s a
+cosmosOn :: Plated a => Fold s a -> Fold s a
```
Comment (phadej) [inline src/Control/Lens/Plated.hs:447]: This is *worse*. It turns rank1types into rank2types. This has worse type-inference properties.

**Comment by sjakobi** [comment]
@emilypi Is this what you had in mind?

@phadej 

> You should use `deepOf`, `deep`, `...` for that. That `Contravariant` is there for a reason, so you cannot make `Traversal` (which would be _very_ easy to break).

To be clear: Are you suggesting that we should prefer `deepOf` over `cosmosOf` in `dhall`?

Could you clarify what changes you're requesting?

Apologies if this is all wrong. I'm rather unfamiliar with `lens`, and I was mostly curious what purpose the `Contravariant` constraint serves. If it should stay, how about documenting why it's there?

**Comment by phadej** [comment]
@sjakobi if you fold, then `cosmosOf` as is should work for you. If you traverse, then you should prefer `deep(Of)`.

**Comment by sjakobi** [comment]
> That `Contravariant` is there for a reason, so you cannot make `Traversal` (which would be _very_ easy to break).

This seems to be the crucial bit that I'd love to understand. I'd appreciate an explanation, if anyone volunteers. :)

Otherwise this seems like a good motivation for me to make more progress in "Optics by Example". ;)

**Comment by phadej** [comment]
If we remove `Contravariant`, and try to use `cosmosOf` as `Setter`:

```haskell
>>> over (cosmosOf plate) (\_xs -> [100 :: Int]) [1..10]
[1,2,3,4,5,6,7,8,9,10]
```

Nothing happens!

What we want to do is maybe

```haskell
> over (deepOf plate _Cons) (\(_x,xs) -> (100, xs)) [1..10 :: Int]
[100,2,3,4,5,6,7,8,9,10]
```

or

```haskell
> transformOf plate (set _head 100) [1..10::Int]
[100,100,100,100,100,100,100,100,100,100]
```

or

```haskell
> rewriteOf plate (\xs -> case xs of (x:xs') | x /= 100 -> Just (100:xs'); _ -> Nothing ) [1..10::Int]
[100,100,100,100,100,100,100,100,100,100]
```

---

I don't have a copy of optics by example, so please tell if their chapter of control.lens.plated is worth recommending.

**Comment by sjakobi** [comment]
Thanks for the great examples, @phadej! :)

So could the "relaxed" `cosmosOf` produce any useful `Setter`s at all? I've played a bit with it but wasn't able to modify the input.

> I don't have a copy of optics by example, so please tell if their chapter of control.lens.plated is worth recommending.

It's pretty short, but a good intro I guess. I'll send you that chapter by email, so you can take a look for yourself.

**Comment by phadej** [comment]
Looks like the chapter has similar examples, but with slightly more prose around them :) Some summary would helped. Some more examples of `uniplate` and `biplate` differences would helped. Though we should have them already in `Data.Data.Lens`.

**Comment by sjakobi** [comment]
I have a hypothesis why `cosmosOf` can't produce useful `Traversal`s: The effects are chained with `*>`, which means that only the return value of the action applied to last child is retained.

Is that right?

Also, does my attempt to document the `Contravariant` constraint make sense?

File: src/Control/Lens/Plated.hs [hunk 2]
```diff
@@ -435,6 +435,9 @@ cosmos = cosmosOf plate
 -- @
 -- 'cosmosOf' :: 'Fold' a a -> 'Fold' a a
 -- @
+
+-- The 'Contravariant' constraint isn't required for the implementation. Since any 'Traversal' produced with 'cosmosOf' is more likely than
```
Comment (phadej) [inline src/Control/Lens/Plated.hs:439]: I think we don't need to publicly (in haddocks) document implementation details. The comment can go under the implementation.

Comment (sjakobi) [reply src/Control/Lens/Plated.hs:439]: The empty line above prevents the comment from being included in the haddocks. I can move it anyway, if you prefer.

Comment (phadej) [reply src/Control/Lens/Plated.hs:439]: empty line is too subtle, it looks like mistake.

Comment (sjakobi) [reply src/Control/Lens/Plated.hs]: Done!

**Comment by sjakobi** [comment]
Thanks @phadej! :) I really appreciate how you take care of so many libraries! :+1:

**Comment by phadej** [comment]
Thanks!

(It's all the same https://hackage.haskell.org/package/category-extras ;) )

**Pull request merged**